### PR TITLE
feature: winui 3 support for Sextant

### PR DESCRIPTION
**Pull request by Gitii**
**What kind of change does this PR introduce?**
Adds support for WinUI 3 (in Sextant).
ReactiveUI already supports WinUI 3

**What is the current behavior?**
WinUI 3 isn't supported in Sextant.

**What is the new behavior?**
WinUI 3 is now supported in Sextant.

**What might this PR break?**
Non-support of WinUI 3.

**Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

This PR is currently in Progress because there are some open questions & problems.
The sample as can be started and played with.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/reactiveui/Sextant?pullRequest=561) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/reactiveui/Sextant?pullRequest=561) before we can accept your contribution.<br/><br/>:x: Gitii [sign now](https://cla.dotnetfoundation.org/reactiveui/Sextant?pullRequest=561)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/reactiveui/Sextant?pullRequest=561) it.</sub>

File: src/Sextant.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29521.150
+# Visual Studio Version 17
+VisualStudioVersion = 17.0.31919.166
```
Comment (Gitii) [inline src/Sextant.sln:4]: VS upgraded silently. Is that a problem?

File: src/Sextant.WinUI/NavigationView.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+            if (viewType is null)
+            {
+                throw new InvalidOperationException(
+                    $"No view could be located for type '{viewModel.GetType().FullName}', contract '{contract}'. Be sure Splat has an appropriate registration.");
+            }
+
+            return viewType;
+        }
+
+        private Page LocatePageFor(object viewModel, string? contract)
```
Comment (Gitii) [inline src/Sextant.WinUI/NavigationView.xaml.cs:376]: Why is there `LocatePageTypeFor` and `LocatePageFor`?
One is used by `PushModal`, the other one by `PushPage`.
Based on the name are supposed to do the same thing, except they aren't.
Different view resolver and `LocatePageFor` has side effects, too.

Can this unified?
Why is there are `RegisterWinUIViewLocator` beside the default one?
The original UWP comments indicate to "return UWP views only" but why?

File: src/Sextant.WinUI/NavigationView.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        /// </summary>
+        public IScheduler BackgroundScheduler { get; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the default back button is visible.
+        /// </summary>
+        public bool IsBackButtonVisible
+        {
+            get => (bool)GetValue(IsBackButtonVisibleProperty);
+            set => SetValue(IsBackButtonVisibleProperty, value);
```
Comment (Gitii) [inline src/Sextant.WinUI/NavigationView.xaml.cs:122]: Exposing the setter is worthless as it's controlled by the component itself.
One might think it's a neat way to hide the default back button, except that it gets turned on again.

File: src/Sextant.WinUI/Mixins/DependencyResolverMixins.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            {
+                throw new ArgumentNullException(nameof(dependencyResolver));
+            }
+
+            if (navigationViewFactory is null)
+            {
+                throw new ArgumentNullException(nameof(navigationViewFactory));
+            }
+
+            Lazy<(TView, ViewStackService)> lazy = new Lazy<(TView, ViewStackService)>(() =>
```
Comment (Gitii) [inline src/Sextant.WinUI/Mixins/DependencyResolverMixins.cs:75]: That's the lazy related change.

File: src/Sextant.WinUI/Mixins/DependencyResolverMixins.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        /// <typeparam name="TView">The view type.</typeparam>
+        /// <typeparam name="TViewModel">The viewmodel type.</typeparam>
+        /// <param name="dependencyResolver">The dependency resolver.</param>
+        /// <param name="contract">The contract.</param>
+        /// <returns>
+        /// The dependencyResolver.
+        /// </returns>
+        public static IMutableDependencyResolver RegisterViewWinUI<TView, TViewModel>(
+            this IMutableDependencyResolver dependencyResolver, string? contract = null)
+            where TView : IViewFor<TViewModel>
```
Comment (Gitii) [inline src/Sextant.WinUI/Mixins/DependencyResolverMixins.cs:128]: The UWP-version has a `new()` constraint here.
Based on the description, shouldn't DI be used to retrieve the instance?
NOTE: I changed this (see line 143).

File: src/Sextant.WinUI/NavigationView.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+
+            return viewType;
+        }
+
+        private Page LocatePageFor(object viewModel, string? contract)
+        {
+            var view = _viewLocator.ResolveView(viewModel, contract);
+            var page = view as Page;
+
+            if (view is null)
```
Comment (Gitii) [inline src/Sextant.WinUI/NavigationView.xaml.cs:381]: The UWP-version also has an duplicated `view is null` check:
```
            if (view is null)
            {
                throw new InvalidOperationException($"Cannot find view for '{viewModel.GetType().FullName}', contract '{contract}' does not implement IViewFor.");
            }
```
this is dead code.
What should it check?
I have removed it.

File: hooks/post-checkout [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
Comment (Gitii) [inline hooks/post-checkout:1]: The hooks have been added automatically.
Is that part of `Sextant's` build process?

File: src/Sextant.sln (hunk 1 above)
Comment (glennawatson) [reply src/Sextant.sln:4]: Vs upgrade is fine. We are using vs2022 by default now.

File: hooks/post-checkout (hunk 7 above)
Comment (worldbeater) [reply hooks/post-checkout:1]: This seems to be Git LFS-related stuff, you can safely remove this stuff to avoid including it into diff I suppose.

Comment (Gitii) [reply hooks/post-checkout]: Thanks for you participation in my PR!
I removed the files and added them to `.gitignore`. It seems to be a bug in git fort windows and VS.

**Comment by Gitii** [comment]
Hi @glennawatson (and @worldbeater if you want to jump in),

I would like to wrap up this PR, but I am unsure how to proceed.
Most of my current open questions won't result in breaking changes but introduce a discontinuity with other platform implementations.  
That's why I would like to leave the code unchanged and just fix the remaining issues:
* UWP support (figuring out why it's broken for me)
* Fix sample (there's a crash for modal routes)

But I would still like to tackle them with follow up discussions and PRs.
There are a few things that are in top of my head right now:
* Simplify build/project setup: Remove conditional compilation and move platforms into separate projects (for example uap/uwp). The current one is neat but caused a lot of confusion until I dug through the project files. That would also affect the nuget packages.
* Refactor `NavigationView`: Move the platform independent code into a separate class (queues, push/pop logic & model view handling). Each platform would provide some "minimal" implementation (mimics what's there today) but offer to easily plug in any "navigational control". 
* Cleanup the extension methods (for example `RegisterViewUWP`). They do provide necessary glue code for the DI but the current implementation is not easy to understand.
* Add more documentation for each of the points above. I had a rough start and had to learn a lot by digging through the codebase. I would also like to look at the samples. Maybe they can be split into distinct parts (showcase specific features or use cases, instead show all of them at the same time).

All of that is outside of the scope of this PR.
First, I want to wrap this one up.
But what's your opinion, ideas and plans about the points that I mentioned?

Best regards,
Gitii

**Comment by Gitii** [comment]
I have decided to close this PR and move the code to a standalone package.
While implementing unit tests I found a few bugs and discovered a several misconceptions in this PR which make it not ready to be merged.

**Pull request closed without merging**